File: algorithms/dynamic programming/max_contiguous_subarray.scala
/**
* Given an array of integers, this algorithm finds the first index, 
* last index and the sum of elements of the largest contigious subarray in linear time.
*/

def max_cont_subarray(a: Array[Int]): (Int,Int,Int) = {
    require(a.nonEmpty)

    var current_start = 0
    var best_start = 0
    var current_end = 0
    var best_end = 0
    var current_sum = a(0)
    var best_sum = a(0)
    var n = a.size
    var i = 1

    while(i < n) {
      if(a(i) > current_sum + a(i)){
        current_sum = a(i)
        current_start = i
      } else {
        current_sum += a(i)
      }
      current_end = i

      if(current_sum > best_sum) {
        best_sum = current_sum
        best_start = current_start
        best_end = current_end
      }
      i += 1
    }

    (best_start, best_end, best_sum)

  }

File: algorithms/dynamic programming/max_contiguous_subarray.cpp
// Given an array of integers,
// find the the largest sum of a contiguous subarray
// O(n) tume, constant space

        // Precompute size of the array
        int n = nums.size();
        if(n == 0) return -1;
        
        // This variable will track the max sum
        // The best we can do for an array of length 1 is just the first element
        int max_sum = nums[0];
        // The subarray we're currently looking at
        int current = nums[0];
        
        // Iterating over all the elements of the input array
        for(int i=1; i<n; i++) {
            // We have two choices:
            // 1. continue the current subarray
            // 2. Start anew with the current element
            current = max(current + nums[i], nums[i]);
            max_sum = max(current, max_sum);
        }
        return max_sum;
    }
};